src/Swarm/Language/Parse.hs

```diff
@@ -286,36 +297,49 @@ mkStmt (Just x) = Binder x
 --     App (App (TDef a) (TDef b)) (TDef x)
 --   This function fix that by converting the Apps into Binds, so that it results in:
 --     Bind a (Bind b (Bind c))
-fixDefMissingSemis :: Term -> Term
+fixDefMissingSemis :: Syntax -> Syntax
 fixDefMissingSemis term =
   case nestedDefs term [] of
     [] -> term
-    defs -> foldr1 (TBind Nothing) defs
+    -- TODO: figure out what should be the Syntax Location of this rewrite
+    defs -> foldr1 (\t1 t2 -> noLoc $ SBind Nothing t1 t2) defs
  where
   nestedDefs term' acc = case term' of
-    def@TDef {} -> def : acc
-    TApp nestedTerm def@TDef {} -> nestedDefs nestedTerm (def : acc)
+    def@(Syntax _ SDef {}) -> def : acc
+    (Syntax _ (SApp nestedTerm def@(Syntax _ SDef {}))) -> nestedDefs nestedTerm (def : acc)
     -- Otherwise returns an empty list to keep the term unchanged
     _ -> []
 
-parseExpr :: Parser Term
+parseExpr :: Parser Syntax
 parseExpr = fixDefMissingSemis <$> makeExprParser parseTermAtom table
  where
   table = snd <$> Map.toDescList tableMap
   tableMap =
     Map.unionsWith
       (++)
-      [ Map.singleton 9 [InfixL (TApp <$ string "")]
+      [ Map.singleton 9 [InfixL (exprLoc2 $ SApp <$ string "")]
       , binOps
       , unOps
-      , Map.singleton 2 [InfixR (TPair <$ symbol ",")]
+      , Map.singleton 2 [InfixR (exprLoc2 $ SPair <$ symbol ",")]
       ]
 
+exprLoc2 :: Parser (Syntax -> Syntax -> Term) -> Parser (Syntax -> Syntax -> Syntax)
```

Comment:
Not so sure about these `exprLoc` functions, it's just adapter for `makeExprParser` usage. I guess we could attach location around the argument parser, but it does not seems useful.